torch/cuda/comm.py

```diff
@@ -120,12 +125,19 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
         inputs, placed on the ``destination`` device.
     """
     output = []
-    itrs = [_take_tensors(tensors, buffer_size) for tensors in inputs]
+    itrs = [_take_tensors(tensors, buffer_size, True) for tensors in inputs]
```

Comment:
It seems to me that weird things can start to happen if different devices produce different configurations of sparse/dense (i.e. buffers will have different sizes and orderings between devices). Also, this function is covered in `is_sparse` conditionals almost entirely. It would be better to scatter lists from different GPUs into sparse/dense buckets (and check that they match in denseness), and do reductions within them independently.